### PR TITLE
pythonPackages.datatable: 0.10.1 -> 0.11.0, fix build

### PR DESCRIPTION
**Pull request by risicle**
###### Motivation for this change
ZHF: #97479

I thought giving `datatable` a bump might have been a short cut to getting it building again - how wrong was I? It looks like it's switched to a new build system and quite a few tweaks needed to be made.

Once we've made some alterations to the source (which doesn't seem to have been shipped in a ready-to-build state), we can use the funny custom build system to build a wheel which can then be installed by the `pipInstallHook`.

Running the tests is again annoying because an ini file setting pytest's `python_files` parameter is missing, meaning we have to inject our own.

Finally I had to fix it up for macos, which luckily only involved adding a manual header path, but meant I could get rid of a bunch of patchery and other cruft.

~Have _not_ tested all subsequent rebuilds, but it's late.~ Now I have and the only failure I get is `python37Packages.optuna` on x86_64 linux (which seems to be a `keras`/`tensorflow` issue?)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/datatable/default.nix [hunk 1]
```diff
@@ -1,45 +1,59 @@
...
-    })
-  ];
+  buildPhase = ''
+    python ci/ext.py wheel
+  '';
 
   propagatedBuildInputs = [ typesentry blessed ];
-  buildInputs = [ llvm ] ++ lib.optionals stdenv.isDarwin [ openmp ];
+  buildInputs = [ llvm pipInstallHook ];
   checkInputs = [ docutils pytest ];
```
Comment (drewrisinger) [inline pkgs/development/python-modules/datatable/default.nix:39]:
```suggestion
  checkInputs = [ docutils pytestCheckHook ];
```

File: pkgs/development/python-modules/datatable/default.nix [hunk 2]
```diff
@@ -1,45 +1,59 @@
...
+    pytestIni = writeText "pytest.ini" ''
+      [pytest]
+      python_files = test_*.py test-*.py
+    '';
+  in ''
+    # skip tests which are irrelevant to our installation or use way too much memory
+    pytest -c ${pytestIni} \
+      -k 'not (test_xfunction_paths or test_fread_from_cmd2 or test_cast_huge_to_str or test_create_large_string_column)' \
+      tests/
   '';
```
Comment (drewrisinger) [inline pkgs/development/python-modules/datatable/default.nix:55]:
```suggestion
  pytestFlagsArray = let
    # ini file required to change python_files setting, not included in tarball
    pytestIni = writeText "pytest.ini" ''
      [pytest]
      python_files = test_*.py test-*.py
    '';
  in [
    "-c ${pytestIni}"
  ];
  disabledTests = [
    # skip tests which are irrelevant to our installation or use way too much memory
    "test_xfunction_paths"
    "test_fread_from_cmd2"
    "test_cast_huge_to_str"
    "test_create_large_string_column"
  ];
```

File: pkgs/development/python-modules/datatable/default.nix [hunk 3]
```diff
@@ -1,45 +1,59 @@
...
 buildPythonPackage rec {
   pname = "datatable";
-  version = "0.10.1";
+  version = "0.11.0";
   disabled = pythonOlder "3.5";
 
   src = fetchPypi {
     inherit pname version;
-    sha256 = "3ce5257c0c4afa96e2b14ca47a0aaf73add195b11de48f4adda50b5ede927436";
+    sha256 = "1fvnval1gy38vx5jab9s5my07nj65msgmn4nwad2xxq03rqh5ihr";
```
Comment (drewrisinger) [inline pkgs/development/python-modules/datatable/default.nix:18]:
```suggestion
    sha256 = "19c602711e00f72e9ae296d8fa742d46da037c2d3a2d254bdf68f817a8da76bb";
```
This is the full sha256 pypi's tarball download page: https://pypi.org/project/datatable/#copy-hash-modal-ba705dff-a3ae-4427-b710-a9742b18f5c9

File: pkgs/development/python-modules/datatable/default.nix [hunk 4]
```diff
@@ -1,45 +1,59 @@
 { stdenv, lib, buildPythonPackage, fetchPypi, substituteAll, pythonOlder
+, pipInstallHook, writeText
 , blessed
 , docutils
 , libcxx
-, libcxxabi
 , llvm
-, openmp
 , pytest
```
Comment (drewrisinger) [inline pkgs/development/python-modules/datatable/default.nix:7]:
```suggestion
, pytestCheckHook
```

**Review by drewrisinger — CHANGES_REQUESTED** [review]
* [x] Diff LGTM minus potentially using ``pytestCheckHook``
* [x] Commits LGTM
* [x] Build in progress, only failure so far is ``python37Packages.cma`` which is broken on ``master``

File: pkgs/development/python-modules/datatable/default.nix (hunk 2 above)
Comment (risicle) [reply pkgs/development/python-modules/datatable/default.nix]: Nice one. Integrated.

File: pkgs/development/python-modules/datatable/default.nix (hunk 3 above)
Comment (risicle) [reply pkgs/development/python-modules/datatable/default.nix]: ok - guess it's good for visibility.

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -1460,7 +1460,7 @@ in {
   datashape = callPackage ../development/python-modules/datashape { };
 
   datatable =
-    callPackage ../development/python-modules/datatable { inherit (pkgs.llvmPackages) openmp libcxx libcxxabi; };
+    callPackage ../development/python-modules/datatable { inherit (pkgs.llvmPackages) libcxx; };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:1463]: I find this more readable:
```
  datatable = callPackage ../development/python-modules/datatable {
    inherit (pkgs.llvmPackages) openmp libcxx libcxxabi; 
  };
```

**Review by jonringer — COMMENTED** [review]
otherwise LGTM

**Review by jonringer — APPROVED** [review]
LGTM

failures are broken on target branch
```
https://github.com/NixOS/nixpkgs/pull/98688
1 package failed to build:
python37Packages.optuna

6 packages built:
python37Packages.dask-xgboost python37Packages.datatable python37Packages.xgboost python38Packages.dask-xgboost python38Packages.datatable python38Packages.xgboost
```

File: pkgs/top-level/python-packages.nix (hunk 5 above)
Comment (risicle) [reply pkgs/top-level/python-packages.nix]: Agreed.

**Pull request merged**